### PR TITLE
Add: Support for NIST NVD CVE Change History API

### PR DESCRIPTION
**Pull request by n-thumann**
## What
This PR adds support for NIST NVD CVE Change History API (see https://nvd.nist.gov/developers/vulnerabilities#divGetCveHistory).

## Why
To query for changes of CVEs.

Notes:
https://csrc.nist.gov/schema/nvd/api/2.0/cve_history_api_json_2.0.schema specifies the `eventName` to be a `string` and not an Enum with predefined values. https://nvd.nist.gov/developers/vulnerabilities#cveHistory-eventName however specifies specific values, so I used an Enum anyways.

## References
VTA-463

## Checklist

- [x] Tests

### CODE REVIEW

File: pontos/nvd/cve_change_history/__init__.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+    main(parser, query_changes)
+
+
+def main(parser: ArgumentParser, func: Callable) -> None:
+    try:
+        args = parser.parse_args()
+        asyncio.run(func(args))
+    except KeyboardInterrupt:
+        pass
```
Comment (bjoernricks) [inline pontos/nvd/cve_change_history/__init__.py:38]:
```suggestion
def parse_args() -> Namespace:
    parser = ArgumentParser()
    parser.add_argument("--token", help="API key to use for querying.")
    parser.add_argument("--cve-id", help="Get history for a specific CVE")
    parser.add_argument(
        "--event-name", help="Get all CVE associated with a specific event name"
    )
    return parser.parse_args()


def main() -> None:
    try:
        args = parse_args()
        asyncio.run(query_changes(args))
    except KeyboardInterrupt:
        pass

if __name__ == "__main__:
    main()
```

File: pontos/nvd/cve_change_history/api.py [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+
+    Should be used as an async context manager.
+
+    Example:
+        .. code-block:: python
+
+            from pontos.nvd.cve_change_history import CVEChangeHistoryApi
+
+            async with CVEChangeHistoryApi() as api:
+                async for cve_change in api.cve_changes(event_name=EventName.INITIAL_ANALYSIS):
```
Comment (bjoernricks) [inline pontos/nvd/cve_change_history/api.py:40]:
```suggestion
                async for cve_change in api.changes(event_name=EventName.INITIAL_ANALYSIS):
```

File: pontos/nvd/cve_change_history/api.py [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+                Default: True.
+        """
+        super().__init__(
+            DEFAULT_NIST_NVD_CVE_HISTORY_URL,
+            token=token,
+            timeout=timeout,
+            rate_limit=rate_limit,
+        )
+
+    async def cve_changes(
```
Comment (bjoernricks) [inline pontos/nvd/cve_change_history/api.py:72]: IMHO it's already obvious that we are querying the CVE changes from the API class name and therefore we can shorten the method name
```suggestion
    async def changes(
```

File: pontos/nvd/cve_change_history/api.py [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+        Returns:
+            An async iterator to iterate over CVEChange model instances
+
+        Example:
+            .. code-block:: python
+
+                from pontos.nvd.cve_change_history import CVEChangeHistoryApi
+
+                async with CVEChangeHistoryApi() as api:
+                    async for cve_change in api.cve_changes(event_name=EventName.INITIAL_ANALYSIS):
```
Comment (bjoernricks) [inline pontos/nvd/cve_change_history/api.py:101]:
```suggestion
                    async for cve_change in api.changes(event_name=EventName.INITIAL_ANALYSIS):
```

File: pontos/nvd/cve_change_history/api.py [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+
+    async def __aexit__(
+        self,
+        exc_type: Optional[Type[BaseException]],
+        exc_value: Optional[BaseException],
+        traceback: Optional[TracebackType],
+    ) -> Optional[bool]:
+        return await super().__aexit__(  # type: ignore
+            exc_type, exc_value, traceback
+        )
```
Comment (bjoernricks) [inline pontos/nvd/cve_change_history/api.py:164]: I don't think you need to override the method.
```suggestion
```

File: pontos/nvd/cve_change_history/api.py [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+                async with CVEChangeHistoryApi() as api:
+                    async for cve_change in api.cve_changes(event_name=EventName.INITIAL_ANALYSIS):
+                        print(cve_change)
+        """
+        total_results: Optional[int] = None
+
+        if bool(change_start_date) ^ bool(change_end_date):
+            raise PontosError(
+                "change_start_date and change_end_date must be provided mutally"
+            )
```
Comment (bjoernricks) [inline pontos/nvd/cve_change_history/api.py:109]: I would love to change that similar to https://github.com/greenbone/pontos/blob/main/pontos/nvd/cve/api.py#L180-L185

File: pontos/nvd/cve_change_history/api.py [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+
+        params: Params = {}
+        if change_start_date and change_end_date:
+            if change_end_date - change_start_date > timedelta(days=120):
+                raise PontosError(
+                    "change_start_date and change_end_date must not be more than 120 days apart"
+                )
```
Comment (bjoernricks) [inline pontos/nvd/cve_change_history/api.py:116]: I am not sure if we should check this. I didn't implement this rule for CVEs and CPEs.

File: pontos/nvd/cve_change_history/api.py [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+        while total_results is None or start_index < total_results:
+            params["startIndex"] = start_index
+
+            if results_per_page is not None:
+                params["resultsPerPage"] = results_per_page
+
+            response = await self._get(params=params)
+            response.raise_for_status()
+
+            data: Dict[str, Union[int, str, Dict[str, Any]]] = response.json(
```
Comment (bjoernricks) [inline pontos/nvd/cve_change_history/api.py:139]: Would be nice to use `JSON` here https://github.com/greenbone/pontos/blob/main/pontos/nvd/api.py#L35
```suggestion
            data: JSON = response.json(
```

**Review by bjoernricks — COMMENTED** [review]
Looks very good already. Could you do me an additional favor and implement the changes from https://github.com/greenbone/pontos/pull/919 where i've added requesting a specific number of results? Should be straight forward.

File: pyproject.toml [hunk 9]
```diff
@@ -111,6 +111,7 @@ pontos-github-actions = 'pontos.github.actions:main'
 pontos-github-script = 'pontos.github.script:main'
 pontos-nvd-cve = 'pontos.nvd.cve:cve_main'
 pontos-nvd-cves = 'pontos.nvd.cve:cves_main'
+pontos-nvd-cve-change-history = 'pontos.nvd.cve_change_history:cve_changes'
```
Comment (bjoernricks) [inline pyproject.toml:114]:
```suggestion
pontos-nvd-cve-change-history = 'pontos.nvd.cve_change_history:main'
```

File: pontos/nvd/models/cve_change.py [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+
+
+@dataclass
+class CVEChange(Model):
+    cve_id: str
+    event_name: EventName
+    cve_change_id: UUID
+    source_identifier: str
+    created: Optional[datetime] = None
+    details: Optional[List[Detail]] = None
```
Comment (bjoernricks) [inline pontos/nvd/models/cve_change.py:46]: Forgot to mention, with requiring Python >= 3.9 we can use [PEP 585](https://peps.python.org/pep-0585/) 
```suggestion
    details: Optional[list[Detail]] = None
```

File: pontos/nvd/cve_change_history/api.py (hunk 7 above)
Comment (n-thumann) [reply pontos/nvd/cve_change_history/api.py:116]: This not checking the date range would result in a bad user experience, IMHO.
E.g. https://services.nvd.nist.gov/rest/json/cvehistory/2.0/?changeStartDate=2022-01-01T00:00:00.000%2B00:00&changeEndDate=2023-01-01T00:00:00.000%2B00:00 just returns a empty `404` response with no information what went wrong, so as a user, I'd be confused what went wrong:
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python3.10/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "/usr/lib/python3.10/asyncio/base_events.py", line 649, in run_until_complete
    return future.result()
  File "<stdin>", line 3, in foo
  File "/home/nthumann/Greenbone/pontos/pontos/nvd/cve_change_history/api.py", line 137, in cve_changes
    response.raise_for_status()
  File "/home/nthumann/Greenbone/pontos/.venv/lib/python3.10/site-packages/httpx/_models.py", line 749, in raise_for_status
    raise HTTPStatusError(message, request=request, response=self)
httpx.HTTPStatusError: Client error '404 ' for url 'https://services.nvd.nist.gov/rest/json/cvehistory/2.0?changeStartDate=2022-01-01T00%3A00%3A00&changeEndDate=2023-01-01T00%3A00%3A00&startIndex=0'
For more information check: https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/404
```
If NVD would return a proper error message, I'd go with that, but sadly they don't, so I IMHO this is the second-best option.

File: pontos/nvd/cve_change_history/api.py (hunk 6 above)
Comment (n-thumann) [reply pontos/nvd/cve_change_history/api.py:109]: I implemented this at first, but then noticed two logical loopholes:
- It's possible that NVD API is called with only the end date set, which is not allowed as per NVD API documentation and results in an empty 404 response, e.g. https://services.nvd.nist.gov/rest/json/cvehistory/2.0/?changeEndDate=2023-01-01T00:00:00.000%2B00:00
- It's possible that the NVD API is called with a date range larger than 120 days, if a users call the method with only the start date and it being over 120 days in the past (see https://github.com/greenbone/pontos/pull/920#discussion_r1387691709). We would try to set the end date to today (if less than 120 days) or to the start date + 120 days, whichever is smaller, but IMHO this hides too much magic from the user and makes the behavior harder to predict.

Comment (n-thumann) [reply pontos/nvd/cve_change_history/api.py]: As discussed via DM, the behavior of the CVE and CPE module should be preserved and calculate an end date automatically. To take the range limit of 120 days into account and to improve user experience by having a similar behavior, both dates will be calculated to use the 120 days range, if not manually specified.

Comment (n-thumann) [reply pontos/nvd/cve_change_history/api.py]: Implemented in https://github.com/greenbone/pontos/pull/920/commits/dfa18773371da9637a2157453ddde7ad919cad56.

File: pontos/nvd/cve_change_history/api.py (hunk 8 above)
Comment (n-thumann) [reply pontos/nvd/cve_change_history/api.py]: Done in https://github.com/greenbone/pontos/pull/920/commits/355bea613c0079a625392e931d15f0f68cb04c9c.

File: pontos/nvd/models/cve_change.py (hunk 10 above)
Comment (n-thumann) [reply pontos/nvd/models/cve_change.py]: Done (+ a few other cases) in https://github.com/greenbone/pontos/pull/920/commits/40903811185bac08ce7fcf8f383e6eafadec0409.

File: pontos/nvd/cve_change_history/api.py (hunk 4 above)
Comment (n-thumann) [reply pontos/nvd/cve_change_history/api.py]: Done in https://github.com/greenbone/pontos/pull/920/commits/1ba86c48e94cdc415e73cc3b9e8f5afcab6086cf.

File: pontos/nvd/cve_change_history/api.py (hunk 3 above)
Comment (n-thumann) [reply pontos/nvd/cve_change_history/api.py]: Done in https://github.com/greenbone/pontos/pull/920/commits/1ba86c48e94cdc415e73cc3b9e8f5afcab6086cf.

File: pontos/nvd/cve_change_history/api.py (hunk 2 above)
Comment (n-thumann) [reply pontos/nvd/cve_change_history/api.py]: Done in https://github.com/greenbone/pontos/pull/920/commits/1ba86c48e94cdc415e73cc3b9e8f5afcab6086cf.

File: pyproject.toml (hunk 9 above)
Comment (n-thumann) [reply pyproject.toml]: Done in https://github.com/greenbone/pontos/pull/920/commits/3ac2f438c24db11f3f8a978e8f60ef8823cb049f.

File: pontos/nvd/cve_change_history/__init__.py (hunk 1 above)
Comment (n-thumann) [reply pontos/nvd/cve_change_history/__init__.py]: Done in https://github.com/greenbone/pontos/pull/920/commits/3ac2f438c24db11f3f8a978e8f60ef8823cb049f.

File: pontos/nvd/cve_changes/api.py [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+            rate_limit=rate_limit,
+        )
+
+    async def changes(
+        self,
+        *,
+        change_start_date: Optional[datetime] = None,
+        change_end_date: Optional[datetime] = None,
+        cve_id: Optional[str] = None,
+        event_name: Optional[EventName] = None,
```
Comment (bjoernricks) [inline pontos/nvd/cve_changes/api.py:79]:
```suggestion
        event_name: Optional[Union[EventName, str]] = None,
```

**Review by bjoernricks — APPROVED** [review]

**Pull request merged**